spock-core/src/main/java/org/spockframework/runtime/PlatformParameterizedSpecRunner.java

```diff
@@ -39,231 +39,38 @@ void runParameterizedFeature(SpockExecutionContext context, ParameterizedFeature
-    for (int i = 0; i < dataProviders.length; i++)
-      try {
-        Iterator<?> iter = GroovyRuntimeUtil.asIterator(dataProviders[i]);
-        if (iter == null) {
-          supervisor.error(context.getErrorInfoCollector(), new ErrorInfo(context.getCurrentFeature().getDataProviders().get(i).getDataProviderMethod(),
-              new SpockExecutionException("Data provider's iterator() method returned null")));
-          return null;
-        }
-        iterators[i] = iter;
-      } catch (Throwable t) {
-        supervisor.error(context.getErrorInfoCollector(), new ErrorInfo(context.getCurrentFeature().getDataProviders().get(i).getDataProviderMethod(), t));
-        return null;
-      }
-
-    return iterators;
-  }
-
-  // -1 => unknown
-  private int estimateNumIterations(SpockExecutionContext context, Object[] dataProviders) {
-    if (context.getErrorInfoCollector().hasErrors()) {
-      return -1;
-    }
-    if (dataProviders.length == 0) {
-      return 1;
-    }
-
-    int result = Integer.MAX_VALUE;
-    for (Object prov : dataProviders) {
-      if (prov instanceof Iterator)
-      // unbelievably, DGM provides a size() method for Iterators,
-      // although it is of course destructive (i.e. it exhausts the Iterator)
-      {
-        continue;
-      }
-
-      Object rawSize = GroovyRuntimeUtil.invokeMethodQuietly(prov, "size");
-      if (!(rawSize instanceof Number)) {
-        continue;
-      }
-
-      int size = ((Number)rawSize).intValue();
-      if (size < 0 || size >= result) {
-        continue;
-      }
-
-      result = size;
+    } catch (InterruptedException ie) {
+      throw ie;
+    } catch (Exception e) {
+      // DataIterator.close doesn't throw
     }
-
-    return result == Integer.MAX_VALUE ? -1 : result;
   }
 
-  private void runIterations(SpockExecutionContext context, ParameterizedFeatureChildExecutor childExecutor, Iterator[] iterators, int estimatedNumIterations) {
-    if (context.getErrorInfoCollector().hasErrors()) {
-      return;
-    }
+  private IIterationRunner createIterationRunner(SpockExecutionContext context, ParameterizedFeatureChildExecutor childExecutor) {
+    return new IIterationRunner() {
+      private final AtomicInteger iterationIndex = new AtomicInteger(0);
 
-    int iterationIndex = 0;
-    while (haveNext(context, iterators)) {
-      IterationInfo iterationInfo = createIterationInfo(context, iterationIndex, nextArgs(context, iterators), estimatedNumIterations);
-      IterationNode iterationNode = new IterationNode(
-          context.getParentId().append("iteration", String.valueOf(iterationIndex++)),
+      @Override
+      public CompletableFuture<ExecutionResult> runIteration(Object[] args) {
+        IterationInfo iterationInfo = createIterationInfo(context, iterationIndex.get(), args, -1);
+        IterationNode iterationNode = new IterationNode(
+          context.getParentId().append("iteration", String.valueOf(iterationIndex.getAndIncrement())),
```

Comment:
🤔 The two calls to `iterationIndex` look like a potential race condition here and should be replaced by a single, incrementing call.

Reply:
done [1c18e7c](https://github.com/spockframework/spock/pull/1479/commits/1c18e7c1e230a08fb207292436d8e892bb04851b)